### PR TITLE
NMS-12165: Make Meta-Data-DSL work with arbitrary values

### PR DESCRIPTION
**Pull request by fooker**
* JIRA (Issue Tracker): http://issues.opennms.org/browse/NMS-12165

### CODE REVIEW

File: core/ipc/rpc/utils/src/main/java/org/opennms/core/rpc/utils/mate/Interpolator.java [hunk 1]
```diff
@@ -52,10 +54,18 @@ private Interpolator() {}
...
-        if (value instanceof String) {
+        if (value == null) {
+            return null;
+        } else if (value instanceof String) {
             return interpolate((String) value, scope);
         } else {
-            return value;
+            // Serialize, interpolate and deserialize the value
+            final Class<?> clazz = value.getClass();
+            final String xml = JaxbUtils.marshal(value);
```
Comment (j-white) [inline core/ipc/rpc/utils/src/main/java/org/opennms/core/rpc/utils/mate/Interpolator.java]: Should we try catching any exceptions that may happen during this process and return the original value if something fails, or leave that up to the caller?

(I suspect the marshaling/unmarshaling may break with some types of objects or in certain circumstances.)

File: core/ipc/rpc/utils/src/test/java/org/opennms/core/rpc/utils/mate/RpcMetaDataUtilsTest.java [hunk 2]
```diff
@@ -36,6 +37,11 @@
 import org.junit.Before;
 import org.junit.Test;
 
+import static org.junit.Assert.assertThat;
+import static org.hamcrest.Matchers.*;
```
Comment (j-white) [inline core/ipc/rpc/utils/src/test/java/org/opennms/core/rpc/utils/mate/RpcMetaDataUtilsTest.java]: Nit: *

File: core/ipc/rpc/utils/src/test/java/org/opennms/core/rpc/utils/mate/RpcMetaDataUtilsTest.java [hunk 3]
```diff
@@ -78,4 +84,33 @@ public void testMetaDataInterpolation() {
...
         Assert.assertEquals("aaa${abc}bbb", interpolatedAttributes.get("attribute10"));
     }
+
+    @Test
+    public void testNonStringMetaDataInterpolation() {
+        final TestObject testObject = new TestObject();
+        testObject.setField("chaos.${ctx1:key1}.example.com");
+
+        final Map<String, Object> attributes = Collections.singletonMap("test-object", testObject);
+        final Map<String, Object> interpolatedAttributes = Interpolator.interpolateObjects(attributes, new MapScope(this.metaData));
```
Comment (j-white) [inline core/ipc/rpc/utils/src/test/java/org/opennms/core/rpc/utils/mate/RpcMetaDataUtilsTest.java]: Can we add a test that specifically covers the PSM objects?

File: core/ipc/rpc/utils/src/main/java/org/opennms/core/rpc/utils/mate/Interpolator.java (hunk 1 above)
Comment (fooker) [reply core/ipc/rpc/utils/src/main/java/org/opennms/core/rpc/utils/mate/Interpolator.java]: God idea. I'm not sure which exceptions to catch? But `Throwable` should do the trick, right?

File: core/ipc/rpc/utils/src/test/java/org/opennms/core/rpc/utils/mate/RpcMetaDataUtilsTest.java (hunk 2 above)
Comment (fooker) [reply core/ipc/rpc/utils/src/test/java/org/opennms/core/rpc/utils/mate/RpcMetaDataUtilsTest.java]: jup

**Comment by fooker** [comment]
I've just realized, that this solution does not allow to put expressions into fields which expect to be non-string types (e.g. the port on the PSM config). Therefore I would like to investigate into other solutions...

**Comment by fooker** [comment]
I've just realized, that this solution does not allow to put expressions into fields which expect to be non-string types (e.g. the port on the PSM config). Therefore I would like to investigate into other solutions...

**Comment by fooker** [comment]
Follow up: https://github.com/OpenNMS/opennms/pull/2708

**Pull request closed without merging**